### PR TITLE
Add `BatchValuesFromIter`

### PR DESCRIPTION
**Pull request by Ten0**
This struct implements `BatchValues` from an `IntoIterator` over things that implement `ValueList`

This is to avoid requiring allocating a new `Vec` containing all the `ValueList`s directly:
with this, one can write:
```rust
session.batch(&batch, BatchValuesFromIter::new(lines_to_insert.iter().map(|l| &l.value_list)))
```
where `lines_to_insert` may also contain e.g. data to pick the statement...

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla-cql/src/frame/value.rs [hunk 1]
```diff
@@ -837,6 +838,70 @@ impl<T: ValueList> BatchValues for Vec<T> {
...
+///
+/// This is to avoid requiring allocating a new `Vec` containing all the `ValueList`s directly:
+/// with this, one can write:
+/// ```ignore
+/// session.batch(&batch, BatchValuesFromIter::new(lines_to_insert.iter().map(|l| &l.value_list)))
+/// ```
+/// where `lines_to_insert` may also contain e.g. data to pick the statement...
+///
+/// The underlying iterator will always be cloned at least once, once to compute the length if it can't be known
+/// in advance, and be re-cloned at every retry.
```
Comment (Ten0) [inline scylla-cql/src/frame/value.rs:851]: Is this true though or is it only serialized once?
Need to check this...

Comment (Ten0) [reply scylla-cql/src/frame/value.rs:851]: Yes it is true, as the `SerializedRequest` is only built in `Connection::send_request`:
https://github.com/scylladb/scylla-rust-driver/blob/014c83a9d55d3beb9ac16a01ddec12cecaabe48c/scylla/src/transport/connection.rs#L684

**Comment by psarna** [comment]
@Ten0 Actually, this pull request made us realize that `write_nth_to_request` is too generic an interface... later, the only caller of this function adds requests one-by-one anyway:
https://github.com/scylladb/scylla-rust-driver/blob/32ef011127f5e31ad922c0e5aa4b714cc05b13fa/scylla-cql/src/frame/request/batch.rs#L57-L60

So I suggest that in order to make this implementation **way** simpler, perhaps `write_to_nth_request` should be rewritten to `append_request`, which does not assume random insertions, but instead only appends to the end? Then you won't need to keep double iterators and the implementation of this PR would become extremely straightforward.

**Comment by Ten0** [comment]
Yeah I had done that only to preserve backwards compatibility, but if we don't care about it that would indeed be a better interface.

That is actually basically what I have implemented in #508 so I'd suggest we just get #508 merged and close this.

**Comment by cvybhu** [comment]
Closing because this has been implemented in #508

**Pull request closed without merging**